### PR TITLE
Add CollationGenerationMessage::Reinitialize

### PR DESCRIPTION
**Pull request by tmpolaczyk**
Allows to change the parachain for which the node will collate.

### CODE REVIEW

**Comment by sandreim** [comment]
Can you detail a bit more why you need this ?

**Comment by girazoki** [comment]
We are working on a project that aims at having collators rotate over parachains they are collating on. Currently what we have found is that once the `collationSubsystem` is initialized, there is no way to tell it to collate on a new chain adaptively.

We did a few tests and this apparently allows us to do so, but we might be missing some things

**Comment by sandreim** [comment]
> We are working on a project that aims at having collators rotate over parachains they are collating on. Currently what we have found is that once the `collationSubsystem` is initialized, there is no way to tell it to collate on a new chain adaptively.
> 
> We did a few tests and this apparently allows us to do so, but we might be missing some things

How is this supposed to work in terms the parachain state given that they all paras would share it ?

**Comment by tmpolaczyk** [comment]
> How is this supposed to work if the parachain state(collator db) is mutated by multiple parachains ?

We will only be collating for a single parachain at a time, and each parachain has a separate db, so that shouldn't be a problem. Unless I'm wrong and there is some other db that we didn't consider.

**Comment by girazoki** [comment]
> > We are working on a project that aims at having collators rotate over parachains they are collating on. Currently what we have found is that once the `collationSubsystem` is initialized, there is no way to tell it to collate on a new chain adaptively.
> > We did a few tests and this apparently allows us to do so, but we might be missing some things
> 
> How is this supposed to work if the parachain state(collator db) is mutated by multiple parachains ?

Also right now I realized we were too optimistic. Right now we will at most work with 2 chains at a time (and orchestrator chain and a collating chain), we have an collator->parachain assignation process that should guarantee you do not move from a chain you are collating unless your chain gets de-registered from the orchestrator chain. At this point in time you do not care if you discard the DB of the de-registered chain.

But long term our vision is that collators dont maintain the state of the system, but rather, that they communicate with a set of nodes we call "state preservers". These preserves do not rotate and **maintain the state for a single chain**.

Long story short, you are still in the very early process of trying things, but seems that been able to adaptively react to stop collating in one chain and start collating in another chain is something we will need.

Is there any easier solution than the one we propose?

**Comment by rphmeier** [comment]
This is probably the best solution to have it handled all in-node. In principle, it shouldn't be a problem for one node to be a collator for many different parachains at the same time.

Rather than re-initializing, it seems more worthwhile to fully adapt the collation-generation and collator-protocol/collator-side to support being a collator for many different parachains at the same time.

Note that there are pending changes to both in #5022 (asynchronous backing feature branch), so probably best to defer changes to when that's merged / make the changes on top of that branch

**Comment by girazoki** [comment]
Thanks @rphmeier, yeah I agree with you. Let us re-evaluate this internally and see if we wait until asynch backing is merged to help adapting the protocol
